add vortex suppressor

hack-wrench opened the pull request:
Добавляет мод к щитам, запрещающий пользоваться вортексом на всем радиусе покрытия

<details>
<summary>Чейнджлог</summary>

```yml
🆑HentaiStorm
rscadd: Добавляет собираемый подавитель вортекса в игру
/🆑
```

</details>

✅ Pull Request полностью завершен, мне не нужна помощь чтобы его закончить.
✅ Я внимательно прочитал все свои изменения и багов в них не нашел.
✅ Я запускал сервер со своими изменениями локально и все протестировал.
✅ Я ознакомился c [Guide to Contribute](https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md).

hack-wrench committed:
add vortex suppressor

hack-wrench committed:
add votex_suppressor mode

hack-wrench committed:
add block outgoing use

hack-wrench committed:
update upkeep use for suppressor

HentaiStorm commented:
Потребление энергии подавителем:
163 KWh = Ядро ИИ (Радиус - 4)
7.8 MWh = Бриг (Радиус - 28)
144 MWh = Станция (Радиус - 120)
650 MWh = Лимит щитов (Радиус - 255)

Doster-d commented:
Было бы круто разрешить путешествовать с вортексом на другие z-левелы, независимо от щитов и когда вортексодебил попытается вернуться на станции, а там стоит блокировка, то он не сможет переместиться.

HentaiStorm commented:
Это фича, что вортекс работает внутрь покрытия, но не из него.

HentaiStorm commented:
Зайти можно, выйти нельзя.

No-Tips commented:
144 Мегаватт? А не дохера ли. Я, конечно, понимаю, что это блокировка вортекса, но сейчас такой выработки достичь фактически невозможно. Мать с коллекторами вплотную вырабатывает до 100Мегаватт, если я не ошибаюсь. То есть, чтобы поддерживать вортекс блокиратор мне нужно будет поставить две матери?

No-Tips commented:
Или я где-то туплю?

Doster-d commented:
> 144 Мегаватт? А не дохера ли. Я, конечно, понимаю, что это блокировка вортекса, но сейчас такой выработки достичь фактически невозможно. Мать с коллекторами вплотную вырабатывает до 100Мегаватт, если я не ошибаюсь. То есть, чтобы поддерживать вортекс блокиратор мне нужно будет поставить две матери?

А что Вы не можете построить коллекторы у генератора гравитации?

HentaiStorm commented:
Оно и рассчитывалось на локальное использование, покрывать ею всю станцию нет смысла. Бриг/Ядро - пожалуйста, имхо конечно, думаю геймдизы подправят где считают нужным, если вообще это нужно.

No-Tips commented:
Гравген вырабатывает 30 беккерелей. Мать вплотную - больше ста.

Doster-d commented:
> Это фича, что вортекс работает внутрь покрытия, но не из него.

Добавь еще, что бы войти было нельзя, если щиты стоят.

No-Tips commented:
> Оно и рассчитывалось на локальное использование, покрывать ею всю станцию нет смысла. Бриг/Ядро - пожалуйста, имхо конечно, думаю геймдизы подправят где считают нужным, если вообще это нужно.

Ну, разве что так. Но опять же, тогда придется пускать условные восемь мегаватт в обход смесов, убивая каждого, кто захочет потрогать провод.

No-Tips commented:
В общем, циферки нужно потвикать, чтобы было реально поставить это.

HentaiStorm commented:
Можно строить локальные материи/стеныурана в отсеках. И кстати, воид шифт ниндзи работает по той же схеме, что и вортекс - войти можно, выйти нельзя.

No-Tips commented:
Эх... Зачем мы тогда Пейна заджоббанили... Он ведь просто готовился к этому пулреквесту...

HentaiStorm commented:
Пейн занимался говноебством, строя их в барах/медбеях.

No-Tips commented:
А может, он хотел в баре\медбее поставить вортекс блокатор, А?

Destrolaric commented:
Предложение, сделать это возможным, только для прокачанных щитов.

MrMagick2104 commented on code/game/objects/items/weapons/teleportation.dm:
```diff
@@ -453,6 +454,18 @@ Frequency:
+			if(T.z == SG_turf.z)
+				if(T.x in SG_turf.x - SG.field_radius to SG_turf.x + SG.field_radius)
+					if(T.y in SG_turf.y - SG.field_radius to SG_turf.y + SG.field_radius)
+						to_chat(usr, SPAN_WARNING("Your manipulator does not work in this area!"))
```
Мне кажется, сообщение в чат выглядит слабовато. Может, лучше кидать малфанкшн при попытке использования, если есть заряд в батарейке?

Destrolaric replied:
Сама строка немного мем, if if if лучше не надо, && намного более приятен глазу.

MrMagick2104 commented on code/modules/shield_generators/modes.dm:
```diff
@@ -73,4 +73,10 @@
+	mode_name = "Vortex Suppressor"
+	mode_desc = "This mode blocks outgoing vortex flow."
+	mode_flag = MODEFLAG_VORTEX_SUPPRESSOR
+	multiplier = 1
```
```suggestion
	multiplier = 1
	hacked_only = 1
```

MrMagick2104 reviewed (commented):
Имхо, стоит не только на вортекс ограничения пилить, а вообще на любой блюспейс (хэндтеле, БоХи, етц.), тогда уж. А то не слишком логично, хотя и по балансу нормально.
Только при емаге бы сделать, либо, хотя бы, только при взломе через проводки, по-моему, лучше.

kcalbCube commented:
> 144 Мегаватт? А не дохера ли. Я, конечно, понимаю, что это блокировка вортекса, но сейчас такой выработки достичь фактически невозможно. Мать с коллекторами вплотную вырабатывает до 100Мегаватт, если я не ошибаюсь. То есть, чтобы поддерживать вортекс блокиратор мне нужно будет поставить две матери?

От твоих слов в гробу пейн начал крутиться и выработал 16 Мегаватт

Rodial-Admiral commented:
Код, конечно, странный.
Ты используешь проверку на радиус:
`if(T.z == SG_turf.z)
if(T.x in SG_turf.x - SG.field_radius to SG_turf.x + SG.field_radius)
if(T.y in SG_turf.y - SG.field_radius to SG_turf.y + SG.field_radius)`
Которая не учитывает:
1) Возможность подрубания щитов в мультизет
2) Hull shielding, в котором подавление телепортации, по идее, вообще не должно работать, поскольку щит не замкнут. 
С механо-точки зрения это смешно еще и потому, что на з-уровне по типу астероида ты можешь выкрутить радиус в 255, врубив халлщилдинг, и наслаждаться вырубленными по всему з-уровню щитами при минимальных энергозатратах. На станции, в принципе, все то же - с определенного момента радиус в режиме халл щилдинга перестает сильно влиять на энергопотребление, потому что станция заканчивается.

По поводу энергопотребления: оно разумно, если исходить из предпосылки, что "подавление вортекса" должно применяться только на небольших радиусах.
Подавлять наивысшую РнДшную технологию дженерик-настройкой с низким энергопотреблением по всему з-уровню было бы странно.

Rodial-Admiral commented:
Логику "впускать, но не выпускать" не понял. Зачем оно тогда надо? Влетел в арсенал/копетонскую вортексом, обратно вылетел хендтелепортером. В целом, получается, что ты можешь влететь ВНУТРЬ защищаемой зоны, где находится генератор, обеспечивающий защиту от вортексов. Который ты можешь выключить, влетев в нее вортексом. 
??????????

HentaiStorm commented:
1. Поправимо.
2. Потребление подавителя всегда постоянно, оно зависит только от радиуса покрытия.
3. Логика "вошел, не вышел" - поправима.

Rodial-Admiral commented:
А. Т.е., по сути получается, что "подавитель" - это не функционал щитов, это, фактически, дополнительный девайс внутри щитов, от самих щитов никак особо не зависящий, кроме как по энергии.
Зачем его тогда вообще делать настройкой щитов, а не отдельным девайсом?

HentaiStorm commented:
Думал над этим вопросом с самого начала, нужно спрайтить отдельный девайс/структуру, окошко настройки. Хотя, думаю можно украсть этот же щит, оставить окно, переписать потребление энерджи, вписать отдельные моды incoming/outgoing flow. Так и сделаю.

hack-wrench committed:
Revert "add vortex suppressor"

This reverts commit 8706339b4f0e8d0c571f497e4208b4ff78c42c31.

hack-wrench committed:
Revert "add votex_suppressor mode"

This reverts commit 30ac8f88c8c3e5dc8a4004ac03250b621ce074b9.

hack-wrench committed:
Revert "add block outgoing use"

This reverts commit 6e3ddf00e1834ea11258e59be96f67865eba45a8.

hack-wrench committed:
Revert "update upkeep use for suppressor"

This reverts commit 3ff8531ee512d96748380e63ffd226522b1df05b.

hack-wrench committed:
update vortex suppressor

Итого:
- Отдельная машина с:
* 15 проводов
* subspace filter
* subspace crystal
* СМЕС-провод
- Дает:
* 3 BLUESPACE, 3 POWER, 3 ENGINEERING
- Необходимо для разблокировки в РнД:
* 4 BLUESPACE, 4 POWER, 4 ENGINEERING
- Имеет: Два мода для блокировки исходящего и входящего вортекса. И также 500 MJ энергии.
- Доступен в карго, в офицерской категории за 100 очков.
- Также блокирует ВОИД ниндзи

HentaiStorm commented:
Итого:
- Отдельная машина с:
* 15 проводов
* subspace filter
* subspace crystal
* СМЕС-провод
- Дает:
* 3 BLUESPACE, 3 POWER, 3 ENGINEERING
- Необходимо для разблокировки в РнД:
* 4 BLUESPACE, 4 POWER, 4 ENGINEERING
- Имеет: Два мода для блокировки исходящего и входящего вортекса. И также 500 MJ энергии.
- Доступен в карго, в офицерской категории за 100 очков.
- Также блокирует ВОИД ниндзи
- Собственный спрайт

hack-wrench committed:
forbidden template

oops

hack-wrench committed:
and .dmi

oop x2

ArmenianOnyx commented:
ПАХПАХАПХ, ну вы зачем так нинзю нагинаете?

kcalbCube commented:
ПХАХАХАХ, заходит такой ниндзя, а там щиты заспидранили. Уже вижу его слезы. И ведь не подьебешь, роль проиграна, суицид только

hack-wrench committed:
forbidden .dme

*with new files

ArmenianOnyx commented:
надо было выкладывать обнову до нерфа вортекса рнд

HentaiStorm commented:
А какой спидран может быть? Залочить станцию - 140 мегаватт. Данным девайсом можно разве что ядро защищать и арсенал (~150 киловатт). Для локального пользования.

kcalbCube commented:
Пейна позвать. Можно успеть кристаллов заказать, пока ниндзя минут 15 в ссдшке

HentaiStorm commented:
Потребление энергии подавителем (С двумя режимами, по отдельности затрачивают лишь половину из списка):
163 KWh = Ядро ИИ (Радиус - 4)
7.8 MWh = Бриг (Радиус - 28)
144 MWh = Станция (Радиус - 120)
650 MWh = Лимит щитов (Радиус - 255)

HentaiStorm commented:
А) Откуда ты знаешь про ниндзю Б) Откуда они знают про его телепорт

HentaiStorm commented:
В) Ниндзя может попросту перехватить подавитель из карго Д) Сломать сервера РнД с чертежом

MalinaGirl commented:
апрув будет только после коммита с инвиз сапресором

HentaiStorm commented:
Хотя, вот наличие его в карго уже намекает на то, что им нужно защищаться от чего-то с телепортом. Считаю метой.

hack-wrench committed:
finally

1) Убран из карго
2) Уравнены необходимые уровни с вортексом (По блюспейсу)

hack-wrench committed:
.dme

forbidden

ArmenianOnyx commented:
> А какой спидран может быть? Залочить станцию - 140 мегаватт. Данным девайсом можно разве что ядро защищать и арсенал (~150 киловатт). Для локального пользования.

— Эй, Пейн, чем мы будем заниматься сегодня вечером? 
— Тем же, чем и всегда, Армениан, захуярим реактор в баре.

No-Tips commented:
Блядь. Только не это. Это можно сделать и без ваших ебанутых материй.

No-Tips commented:
И такой щит можно заспидранить за 15 минут вдвоем.

MrMagick2104 commented:
Я бы предложил запилить запрет для телепортера мехов (если ещё не) и обычного хэндтеле, а потом данную машинерию в блюспейс суппрессор переименовать. Ибо хэндтеле смешной тоже бывает.
И "вортекс суппрессор" - не звучит, с логической точки зрения, по-моему. Они же все на магическом блюспейсе работают, как-никак, чего бы этому суппрессору всё остальное не блокировать?
Либо вообще впилить настройки того, что блокировать, а что нет, с увеличением потребления энергии соответственно уровню блокируемого предмета.

ArmenianOnyx commented:
да кстати, пусть и все другие виды телепортов блокирует. А это пр выглядит чисто ради поржать над нинзей

Rodial-Admiral commented:
Вортекс-манипулятор работает с вортексом, а не с блуспейсом. Не надо тут фандомы путать!
Я ему тут артронную энергию почти закодил, да..

А по теме - каргоящики действительно не нужны были. 
Что касается "блуспейс-подавителя" - почему бы не назвать приблуду "пространственным стабилизатором"? Это бы объяснило блокировку принципиально разных видов телепортов, и меньше проблем было бы с метой, поскольку открывается-то оно раньше, чем воркекс.

Doster-d commented:
Ого, ого, да тут у нас есть ЯРСдупнутый юмопереможный рагулевортанутый стабилизационный пескарь!

MrMagick2104 commented:
> пространственным стабилизатором

Предлагаю также альтернативное название - "Якорь реальности".

Doster-d commented:
> > пространственным стабилизатором
>
> Предлагаю также альтернативное название - "Якорь реальности".

Так уже использовано в ЯРС.

kreeperHLC commented:
> > пространственным стабилизатором
> 
> Предлагаю также альтернативное название - "Якорь реальности".

А область в которой он находится будет подгружена даже когда игрок уйдет от неё?

HentaiStorm commented:
> объяснило блокировку принципиально разных видов телепортов, и меньше проблем было бы с метой, поскольку открывается-то оно раньше, чем воркекс.
В данный момент он открывается одинаково с вортексом.

HentaiStorm commented:
>да кстати, пусть и все другие виды телепортов блокирует. А это пр выглядит чисто ради поржать над нинзей

А) Эта приблуда изучается одновременно с вортексом Б) У нее пиздецкое энергопотребление В) Ей нужны компоненты из шахтерских ресурсов
Я уже не говорю о том, что ниндзя может спокойно вычтать каждый из пунктов украв ресурсы/сломав сервер/инженерку или банально перехватив плату с подавителем.

Rodial-Admiral commented:
> в данный момент он открывается одинаково с вортексом

Требования по техам для машины из твоего кода: 
req_tech = list(TECH_BLUESPACE = 5, TECH_POWER = 5, TECH_ENGINEERING = 5)

Требования по техам для вортекса: 
req_tech = list(TECH_MATERIAL = 8, **TECH_BLUESPACE = 5**, TECH_MAGNET = 7, **TECH_POWER = 8**, TECH_DATA = 5, **TECH_ENGINEERING = 8**)
Жирным выделил то, что присутствует в машине - совпадает только блуспесс. Плюс еще сверху накинуты три других теха.

Мне кажется, /немного/ неодинаково.

HentaiStorm commented:
Суть то в блуспейсе

HentaiStorm commented:
Что у него один с вортексом уровень по голубому.

MrMagick2104 commented:
Блюспейс 5 можно и без ресурсов вкачать же.
> сломав сервер

Ломание сервера к стиранию уже напечатанных плат и уровней из консоли не приводит.

> ниндзя может спокойно вычесть каждый

Если ниндзя превентивно всю научку не вырезает вместе с СБ, то, в большинстве случаев, не может. Ему не приходит СМС от клана Паука, или на кого он там работает, сообщающая о том, что где-то строят или собираются построить (изучили) ЯРС.

kcalbCube commented on code/modules/vortex_suppressor/obj/machinery.dm:
```diff
@@ -0,0 +1,251 @@
+	if(!running)
+		return
+
+	upkeep_power_usage += round(((((suppressor_radius * 2) ** 2) / 2) * ENERGY_UPKEEP_SUPPRESSOR * 50) * upkeep_multiplier)
```
```suggestion
	upkeep_power_usage += round((suppressor_radius * 2) ** 2 * ENERGY_UPKEEP_SUPPRESSOR * 25 * upkeep_multiplier)
```

kcalbCube commented on code/modules/vortex_suppressor/obj/machinery.dm:
```diff
@@ -0,0 +1,251 @@
+	if(current_energy <= 0)
+		energy_failure()
+
+/obj/machinery/power/vortex_suppressor/attackby(obj/item/O as obj, mob/user as mob)
```
Убери

HentaiStorm commented:
> Блюспейс 5 можно и без ресурсов вкачать же.
Но для постройки вортекса требуются ресурсные компоненты

> Ломание сервера к стиранию уже напечатанных плат и уровней из консоли не приводит.
Отвертка + лом по серверу, в консоли Settings ->Reset RD databse (если не ошибаюсь)

MrMagick2104 commented:
> Отвертка + лом по серверу, в консоли Settings ->Reset RD databse (если не ошибаюсь)

Это ресетает данные сервера. Больше оно ничего не ресетает, разве нет?

HentaiStorm commented:
Конкретно оно ресетит данные консоли. Ресетнуть сервер невозможно из-за древнего гриф протекта (который по хорошему бы спилить).

hack-wrench committed:
Merge branch 'dev' into patch-19

Vallat commented on code/modules/vortex_suppressor/obj/machinery.dm:
```diff
@@ -0,0 +1,251 @@
+
+
+/obj/machinery/power/vortex_suppressor/attack_hand(mob/user)
+	ui_interact(user)
```
Открываем интерфейс и, если вскрыты провода, открываем еще и панельку с ними? Не лучше ли разделить эти два процесса? Например, так
```
    if(!panel_open)
        ui_interact(user)
    else
        wires.Interact(user)
```

MrMagick2104 replied:
Разве типичная машинерия (вендоматы, те же щиты, другие штуки) не действует по такой логике одновременного открытия?
По-моему, как раз таки действует, для постоянства менять не стоит, имхо.



Ну и банально удобнее, если сразу оба меню открываются, т.к. иначе придётся закручивать панель, если случайно нажал на крестик или нужно обновить главное меню.

Vallat replied:
> Разве типичная машинерия (вендоматы, те же щиты, другие штуки) не действует по такой логике одновременного открытия?
> По-моему, как раз таки действует, для постоянства менять не стоит, имхо.

Если так, то ладно. С логикой того как работает игра пусть геймдизайнеры разбираются

Vallat commented on code/modules/vortex_suppressor/obj/machinery.dm:
```diff
@@ -0,0 +1,251 @@
+		icon_state = "generator0"
+
+
+/obj/machinery/power/vortex_suppressor/New()
```
```suggestion
/obj/machinery/power/vortex_suppressor/Initialize()
```

Vallat commented on code/modules/vortex_suppressor/obj/machinery.dm:
```diff
@@ -0,0 +1,251 @@
+
+
+/obj/machinery/power/vortex_suppressor/New()
+	..()
```
```suggestion
	. = ..()
```

Vallat commented on code/modules/vortex_suppressor/obj/turf.dm:
```diff
@@ -0,0 +1,16 @@
+// Teleport Suppressor
+/turf/proc/check_unallow_suppressor_flag(var/flag, var/warning_message)
+	for(var/obj/machinery/power/vortex_suppressor/VS in world)
```
Ужасно, пожалуйста, никогда не делай так. Если уж сильно не хочется создавать список с нужными тебе объектами, есть же лист со всей машинерией
```suggestion
	for(var/obj/machinery/power/vortex_suppressor/VS in SSmachines.machinery)
```
Особенно учитывая то, как часто этот процесс будет вызываться, искать объекты по всему миру - смерть в лучах лагов.

Vallat reviewed (changes requested):
Исправить пару мелочей для улучшения читаемости, из критичного - цикл по всему миру

TobyThorne commented:
Так и не понял, в чём смысл. Огромная куча когда чтобы контрить один-единственный вортекс. Он и так уже сильно занёрфлен, если сравнивать с первоначальным видом.

PlinHost commented:
> Так и не понял, в чём смысл. Огромная куча когда чтобы контрить один-единственный вортекс. Он и так уже сильно занёрфлен, если сравнивать с первоначальным видом.

Выше вон предлагают заблокировать все телепорты и блюспейс, тогда вортекс будет не единственный.

hack-wrench committed:
Merge branch 'ChaoticOnyx:dev' into patch-19

hack-wrench committed:
finally

hack-wrench committed:
linter fix

hack-wrench commented:
Завалялся патч к фиче годовалой давности. Вроде готово.

hack-wrench commented:
Но только для вортекса, когда-то в планах было еще отключить работу БоХов, но руки так и не дошли. Думаю вы и сами закодить сможете? Там проки подавителей по турфам.

Doster-d reviewed (approved):
Круто.

kreeperHLC commented:
Почти год прошёл, появились конфликты
